### PR TITLE
Icon button

### PR DESCRIPTION
**Pull request by yasintz**
son resim

### CODE REVIEW

File: src/components/ui/button.js [hunk 1]
```diff
@@ -1,63 +1,117 @@
...
     const {
-      disabled, size, groupOrder, styleType,
+      disabled, size, groupOrder, styleType, IconOnly, loading,
     } = this.props;
 
-    return cls(style.common, style.groupOrder[groupOrder],
-      style.label[size], style.container[styleType], {
+    return cls(
+      this.style.common,
+      this.style.center,
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: seems like you apply center all the time. we can put those classes in styles.common

File: src/css/vendor/tailwind.setup.js [hunk 2]
```diff
@@ -53,15 +52,16 @@ const colors = {
   xdarkblue: '#192344',
   black: '#383b41',
   gray: '#919EAB',
+  gunmetal: '#454F5B',
   lgray: '#C4CDD5',
   xlgray: '#F9FAFB',
+  palegray: '#f4f6f8',
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind.setup.js]: where did you get the name? xlgray is really close to this value, cant we use that instead?

File: src/css/vendor/tailwind.setup.js [hunk 3]
```diff
@@ -53,15 +52,16 @@ const colors = {
   xdarkblue: '#192344',
   black: '#383b41',
   gray: '#919EAB',
+  gunmetal: '#454F5B',
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind.setup.js]: lets call this `dgray` instead

File: test/components/ui/button.test.js [hunk 4]
```diff
@@ -20,8 +20,4 @@ describe('UI Component: <Button />', () => {
     button.simulate('click');
     expect(onClickSpy.callCount).toEqual(1);
   });
-
-  test('has text', () => {
-    expect(button.text()).toBe('button-text');
```
Comment (SelmanKahya) [inline test/components/ui/button.test.js]: why did you remove this test? can you instead update it to pass? you can verify if icons are rendered.

File: src/components/ui/button.js [hunk 5]
```diff
@@ -1,63 +1,117 @@
...
+      middle: 'rounded-none',
+      none: 'rounded-lg',
+    },
+    icon: {
+      large: 'w-12 h-12',
+      medium: 'w-10 h-10',
+      small: 'w-8 h-8',
+    },
+    label: {
+      large: 'w-32 py-3 text-lg',
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: you shouldnt give width or height to labels, instead you should give paddings.

File: src/components/ui/button.js [hunk 6]
```diff
@@ -1,63 +1,117 @@
...
+      medium: 'w-28 py-2 text-base leading-base',
+      small: 'w-24 py-2 text-sm',
+    },
+    textIcon: {
+      large: '20%',
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: what are these numbers?

File: src/components/ui/button.js [hunk 7]
```diff
@@ -1,63 +1,117 @@
...
-    small: 'px-6 py-2 text-sm',
-  },
-};
+import { Loader } from 'react-feather';
 
 class UIButton extends React.Component {
+  style = {
+    center: 'flex justify-center items-center',
+    common:
+      'button inline-block text-center cursor-pointer border border-transparent border-box active:opacity-100',
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: why do you need active:opacity-100?

File: src/components/ui/button.js [hunk 8]
```diff
@@ -1,63 +1,117 @@
...
-    middle: 'rounded-none',
-    none: 'rounded-lg',
-  },
-  label: {
-    large: 'px-8 py-3 text-lg',
-    medium: 'px-6 py-2 text-base leading-base',
-    small: 'px-6 py-2 text-sm',
-  },
-};
+import { Loader } from 'react-feather';
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: do not use icon directly, instead import our icon component.

File: src/components/ui/button.js [hunk 9]
```diff
@@ -1,63 +1,117 @@
...
+      type,
+      disabled,
+      IconLeft,
+      IconOnly,
+      IconRight,
+      loading,
+      size,
     } = this.props;
-
+    const LeftIcon = () => (IconLeft ? (
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: this is extremely inefficient. you should not create react objects in your render method. render should be plain and simply, and if state/props don't change, separate calls to render should always return the same value. senin durumunda her render cagirdiginda yeni iconlar uretiyorsun, bir onceki render() output'undan farkli oldugu icin ekrana tekrar ciziliyor, performans acisindan inanilmaz problemli ve react'e ters bir pattern.

File: src/components/ui/button.js [hunk 10]
```diff
@@ -1,63 +1,117 @@
...
+      </React.Fragment>
+    ) : null);
+    const TextButton = () => (
+      <React.Fragment>
+        <RightIcon />
+        {label}
+        <LeftIcon />
+      </React.Fragment>
+    );
+    const Loading = () => <Loader size={IconOnly ? '60%' : this.style.textIcon[size]} />;
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: instead of using Loader directly, use our Icon component

Comment (SelmanKahya) [reply src/components/ui/button.js]: neden 20-30 gibi degerler yerine yuzde kullanmaliyiz?

**Comment by SelmanKahya** [comment]
thanks for the screenshot, looks good overall, left some comments

File: src/components/ui/button.js (hunk 9 above)
Comment (yasintazeoglu) [reply src/components/ui/button.js]: i did it because it was done before  `components/ui/input.js`

File: src/components/ui/button.js (hunk 10 above)
Comment (yasintazeoglu) [reply src/components/ui/button.js]: ekrani buyutup kuculltugumuzde butonlarda buna bagli olarak buyuyup kuculmesi icin

File: src/components/ui/button.js (hunk 1 above)
Comment (yasintz) [reply src/components/ui/button.js]: > seems like you apply center all the time. we can put those classes in styles.common

File: src/components/ui/button.js [hunk 11]
```diff
@@ -1,22 +1,21 @@
...
 const style = {
-  common: 'button inline-block text-center cursor-pointer border border-transparent border-box',
-  container: {
-    danger: 'text-white bg-red',
-    outline: 'text-blueyGrey hover:bg-paleGrey border-lightBlueGrey',
-    plain: 'text-primary hover:text-primaryDark bg-transparent',
-    primary: 'text-white bg-primary',
-    secondary: 'text-white bg-blueyGrey',
-  },
+  common: 'button text-center border-2 border-box active:opacity-100 focus:outline-none',
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: why did you delete inline-block, and add border?

File: src/css/vendor/tailwind_source.css [hunk 12]
```diff
@@ -20,13 +20,16 @@
 .theme-light {
     --primary : #325FFF;
     --primaryLight: #0f97ff;
-    --primaryDark: #212B36;
+    --primaryDark: #243983;
+    --primaryDarkx: #182fff;
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind_source.css]: what is primaryDarkx?

File: src/css/vendor/tailwind_source.css [hunk 13]
```diff
@@ -20,13 +20,16 @@
...
+    --dark: #212B36;
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind_source.css]: did okan provide you with the name `dark`?

File: src/css/vendor/tailwind_source.css [hunk 14]
```diff
@@ -37,6 +40,7 @@
     --primary : #ff3e3e;
     --primaryLight: #0f97ff;
     --primaryDark: #ffffff;
+    --dark: #212B36;
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind_source.css]: every color defined above should have corresponding black theme color.

yukarida yer alan her rengin burada bir eslenigi olmali, lightpalegrey vs eklenmesi lazim buraya da

File: test/components/ui/button.test.js [hunk 15]
```diff
@@ -1,27 +1,49 @@
 import React from 'react';
-import { shallow } from 'enzyme';
+import { Camera, Loader } from 'react-feather';
```
Comment (SelmanKahya) [inline test/components/ui/button.test.js]: dont import from react-feather, use `ui/icon.js`

File: test/components/ui/button.test.js [hunk 16]
```diff
@@ -1,27 +1,49 @@
...
+import { mount } from 'enzyme';
```
Comment (SelmanKahya) [inline test/components/ui/button.test.js]: why mount, not shallow?

File: test/components/ui/button.test.js [hunk 17]
```diff
@@ -1,27 +1,49 @@
...
 
-  test('renders button', () => {
+  test('renders button correctly', () => {
     expect(button.length).toEqual(1);
+    expect(buttonIcon.length).toEqual(2);
+    expect(buttonLabel.length).toEqual(1);
+    expect(buttonLabel.text()).toEqual('button-text');
+  });
+  test('renders button  props', () => {
+    expect(button.props().disabled).toEqual(false);
```
Comment (SelmanKahya) [inline test/components/ui/button.test.js]: boyle bir teste gerek yok, button render edip props'u kontrol ediyorsun, testing fail olma ihtimali yok.

File: src/components/ui/icon-button.js [hunk 18]
```diff
@@ -0,0 +1,101 @@
+import React from 'react';
```
Comment (SelmanKahya) [inline src/components/ui/icon-button.js]: this is LOTS OF duplication, we should use button.js, cant you refactor it so that we dont need to create a file that is 95% similar?

File: src/components/ui/button.js [hunk 19]
```diff
@@ -66,6 +126,7 @@ class UIButton extends React.Component {
 UIButton.defaultProps = {
   disabled: false,
   groupOrder: 'none',
+  onClick() {},
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: no need to add this

File: src/components/ui/button.js [hunk 20]
```diff
@@ -27,37 +26,98 @@ const style = {
...
+      disabled, size, groupOrder, styleType, loading,
     } = this.props;
+    const container = {
+      danger: [
+        'text-white bg-red border-transparent',
+        {
+          'focus:border-dark': !loading || !disabled,
+        },
+      ],
+      outline: [
```
Comment (SelmanKahya) [inline src/components/ui/button.js]:
```
outline: {
  'text-blueyGrey bg-white border-lightBlueGrey': true,
  'hover:bg-lightPaleGrey focus:border-blueyGrey active:shadow-button-blueyGrey active:bg-paleGrey':
            !loading || !disabled,
}
```
is easier to understand

File: src/components/ui/button.js (hunk 11 above)
Comment (yasintz) [reply src/components/ui/button.js]: inline-block a gerek yoktu buttona sabit bir genislik ve yukseklik vermiyoruz paddinglerle hallediyoruz

Comment (yasintz) [reply src/components/ui/button.js]: border vermemin sebebi active ve focus olduklarinda borderlarinda degisim oluyor button olusurken border verip onu assagida gerekli buttonlari icin transparent yapiyorum

File: src/css/vendor/tailwind_source.css (hunk 12 above)
Comment (yasintz) [reply src/css/vendor/tailwind_source.css]: buttonlarin hover ve focus renleri colors paletinin icinde yok @okandungel  ile konustum eklememi soyledi diger renklerle halletmeye calistim ama iyi durmadi isimlerini bu seklide verdim o sonra duzeltiriz dedi

File: src/css/vendor/tailwind_source.css (hunk 13 above)
Comment (yasintz) [reply src/css/vendor/tailwind_source.css]: colors icinde var dark https://app.zeplin.io/project/5be16ddeab2d281f31bba667/screen/5bf4707f7b5f913ebc721e14

File: src/css/vendor/tailwind.setup.js [hunk 21]
```diff
@@ -486,8 +487,14 @@ module.exports = {
   */
 
   minHeight: {
+    4: '4rem',
+    6: '6rem',
+    8: '8rem',
+    10: '10rem',
+    12: '12rem',
+    32: '8rem',
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind.setup.js]: bu dogru, usttekiler yanlis???

ornegin 4 birim 4x4=16 px yani 1 rem olmali, 4 rem yaziyor

File: src/css/vendor/tailwind.setup.js [hunk 22]
```diff
@@ -469,7 +464,13 @@ module.exports = {
   |
   */
 
-  minWidth: {},
+  minWidth: {
+    4: '4rem',
+    6: '6rem',
+    8: '8rem',
```
Comment (SelmanKahya) [inline src/css/vendor/tailwind.setup.js]: birimler yanlis

File: src/components/ui/button.js [hunk 23]
```diff
@@ -1,67 +1,153 @@
...
-        'border-l-0': groupOrder === 'first' && ['outline'].indexOf(styleType) < 0,
-        // remove border-right on last item except outline styleType
-        'border-r-0': groupOrder === 'last' && ['outline'].indexOf(styleType) < 0,
-        // remove border-top-bottom and add bordercolor except outline styleType
-        'border-transparent20 border-t-0 border-b-0':
-          groupOrder !== 'none' && ['outline'].indexOf(styleType) < 0,
-        'cursor-not-allowed': disabled,
-        'hover:opacity-80': !disabled && ['primary', 'secondary'].indexOf(styleType) > -1,
-        'opacity-20': disabled && styleType !== 'plain',
-        'text-lightBlueGrey hover:text-lightBlueGrey': disabled && styleType === 'plain',
```
Comment (SelmanKahya) [inline src/components/ui/button.js]: bunu silip yerine koymamissin, bug a sebep olacak gibi gorunuyor

bu gibi es gecilem birsey olmadigina emin olabilir miyiz?

Comment (yasintz) [reply src/components/ui/button.js]: buna zaten gerek yok https://app.zeplin.io/project/5be16ddeab2d281f31bba667/screen/5be16e0732cfe523b295113e burda butun buttonlarin disabled hallerinde text-white

Comment (yasintz) [reply src/components/ui/button.js]: plainde ise disabledde hic bir sey degismiyor
